### PR TITLE
Improve printing of reverse coercions

### PR DESCRIPTION
**Pull request by proux01**
When a a term [x] is elaborated into [x'] through a reverse coercion, return the term [rev_coerce x' x] that is convertible to [x'] but printed [x], thanks to the coercion [rev_coerce]. Previously, the returned term was directly [x'].

- [x] Added / updated **test-suite**.

- [ ] Added **changelog**.
- [x] Added / updated **documentation**.
  
  - [ ] ~Documented any new / changed **user messages**.~
  - [ ] ~Updated **documented syntax** by running `make doc_gram_rsts`.~

- [ ] Opened **overlay** pull requests.

<!-- Pointers to relevant developer documentation:

Contributing guide: https://github.com/coq/coq/blob/master/CONTRIBUTING.md

Test-suite: https://github.com/coq/coq/blob/master/test-suite/README.md

Changelog: https://github.com/coq/coq/blob/master/doc/changelog/README.md

Building the doc: https://github.com/coq/coq/blob/master/doc/README.md
Sphinx: https://github.com/coq/coq/blob/master/doc/sphinx/README.rst
doc_gram: https://github.com/coq/coq/blob/master/doc/tools/docgram/README.md

Overlays: https://github.com/coq/coq/blob/master/dev/ci/user-overlays/README.md

### CODE REVIEW

File: pretyping/coercion.ml
Comment (SkySkimmer) [inline pretyping/coercion.ml:654]: Perhaps ReplaceCoe could be extended so that reapply_coercions also inserts the rev_coerce?

**Comment by CohenCyril** [comment]
Fantastic. The only thing that worries me is the universe polymorphic status of the constant `rev_coerce`. It's likely it's used everywhere at various universe levels, so it would make sense to make sure it is universe polymorphic. On the other hand since the main and only current use of this is in an experimental HB/mathcomp branch and HB does not support universe polymorphism yet, it would be premature to make `rev_coerce` universe polymorphic (unless it already works out of the box, because it might not interact too much with HB?).
- @gares what's your take on it?
- @proux01 would you mind trying making it universe polymorphic?

Nitipicking: since `rev_coerce` is neither input by the user, nor displayed, I would advise to give it a longer, more "english" name, like `reverse_coercion`.

**Comment by gares** [comment]
A constant, x or x', is always either upoly or not.
We could have 2 reverse_coercion, one of which is upoly.
Moreover, if you don't need upoly you don't want to have an upoly constant around. Maybe I'm just too conservative, but it used to be the case upoly code was more buggy (eg tactics), so I'd rather play safe.

**Comment by proux01** [comment]
@CohenCyril you're right, done. (there was only one `HB.instance` not going through in MathComp, but it was easy to workaround (c.f. https://github.com/math-comp/math-comp/commit/9f0ffef29f84dcfdbf61fd517bed9aea841396f3) so let's go for it)

**Comment by proux01** [comment]
@gares do you mean that universe monomorphic constants could'nt cause universe issues with a monomorphic `reverse_coercion`? I'm not convinced that two different monomorphic constants with different universe levels couldn't cause a conflict by setting additional universe constraints with a monomorphic `reverse_coercion` (which in my understanding is what @CohenCyril (and I) fear).

**Comment by gares** [comment]
I think you are right, but then this approach is a bit scary ;-)

I wonder if an uglier encoding would be better, eg `.. x ..` -> `.. (let original_term : T := x in x') ..` and have a special rule for this degenerate letin which uses a recognizable ident.

**Comment by proux01** [comment]
Sounds good except that `Eval simpl in (let _original_term := x in x')` yields `x'` which means the nice printing would disappear at the first simplification. Maybe not what we want?

**Comment by Janno** [comment]
I am very confused by this change but I have no experience with reversible coercions. In any case, instead of changing what is printed this PR seems to change the inferred term instead. Is my understanding correct?

In that case this seems like a very dangerous thing to do. Syntactic pattern matching will have to account for the newly introduced term. Another (related) problem that comes to mind immediately is that this will either slow down or break TC search on classes whose arguments are inferred this way. (Again, I haven't really played with reversible coercions so this might not be possible?) Instead of `MyClass my_inferred_structure my_value` the inferred term could be `MyClass (rev_coerce my_inferred_structure my_original_structured) my_value`. If the TC database is transparent by default and `my_structure` is hint opaque then the search can no longer discriminate based on `my_structure` and will likely yield too many applicable (but ultimately unsuccessful) hints. If the database is opaque by default the search will likely fail entirely unless the hints similarly have `rev_coerce` in them, which I don't think is always going to be the case.

> On the other hand since the main and only current use of this is in an experimental HB/mathcomp branch

I am not sure what "of this" refers to exactly. If it refers to reversible coercions I know that they already showed up in unexpected (but not unwelcome) places in the form of terms that typecheck in 8.16 when they hadn't before. So they are definitely used in the wild even if people don't necessarily notice.

Sorry if I misunderstood what is going on. I just want to avoid being even more surprised in future coq by what terms *really* look like compared to what is printed.

**Comment by proux01** [comment]
> I'm just scared that using upoly will open a can of worms.

Well, that might be a good opportunity to fix many things. (just trying to look at the bright side of it, I perfectly agree with you)

**Comment by andres-erbsen** [comment]
While I would be very excited to have more concise printing of elaborated code, I think the approach to achieve this needs more discussion. In short, I share the concerns @Janno posted above.

More generally, this PR seems to be based on a unknown-to-me vision of how elaboration and tactic programming would interact, so I have a hard time understanding it. Specifically, I don't see what use coercions would be if using them produced different terms than writing out what you meant in painful detail, because I don't expect most tactics to work on the coercion-derived code. I am getting the sense that the PR authors would consider this acceptable and expected, but I would consider any instance where printing (or anything else) depends on whether the term was originally elaborated using coercions or without to be a bug (and considering "the term" to include a constant specific to the coercion mechanism would be deeply dissatisfying). Concretely, would `congruence` (and other tactics) need to be updated to interpret (ignore) the new constant to retain their power?

> Syntactic pattern matching will have to account for the newly introduced term. Another (related) problem that comes to mind immediately is that this will either slow down or break TC search on classes whose arguments are inferred this way

IIUC, anything like this would be quite costly to support in fiat-cryptography tactics, so if that's the only available behavior, the features affected by this change would be broken as far as I'm concerned. This may not be an immediate practical concern as I am not sure whether we use reversible coercions at all.

**Comment by proux01** [comment]
> [...] whether the term was originally elaborated using coercions or without [...] Concretely, would `congruence` (and other tactics) need to be updated to interpret (ignore) the new constant to retain their power?

No: reverse coercions are only a way to give a meaning to term that did not typecheck before their introduction. So any tactic will "retain its power" in the sense it should keep handling anything it previously handled in the same way (assuming it is not calling the elaborator and handling its failures in a specific way but most tactic work on already elaborated terms).

> Syntactic pattern matching will have to account for the newly introduced term.

Of course, if you start using reverse coercions and feeding this new terms to your own specific tactics, some adptation might be needed to handle the new terms but that's life. Again, this should only be about the new terms, backward compatibility should hold.

> Another (related) problem that comes to mind immediately is that this will either slow down or break TC search on classes whose arguments are inferred this way

Here things are likely to work out of the bow as typeclass search will simply unfold the identity `reverse_coercion`.

> I am not sure whether we use reversible coercions at all.

Basically, if your code compiles with any Coq < 8.16, it means you are not.

**Comment by Janno** [comment]
> > Another (related) problem that comes to mind immediately is that this will either slow down or break TC search on classes whose arguments are inferred this way
> 
> Here things are likely to work out of the bow as typeclass search will simply unfold the identity `reverse_coercion`.

I suspect you allude to this with "likely" but I'd like to make this a bit more concrete for the benefit of people who do not work with typeclasses a lot: This is only true for hint databases that treat constants as transparent by default. Granted, that is often the case because the opposite behavior makes TC search infeasible outside of very specific and restrictive scenarios. However, when performance is a concern, the only way I know to use TC search effectively and efficiently is to use a transparent-by-default database and to make almost all definitions TC opaque explicitly to benefit from the discrimination nets. Failing hint applications can be very, very expensive and even if every individual one is cheap they add up incredibly quickly when hints look too applicable to the discrimination net. In the development I work on a single missed opacity annotation can make a proof 25%, 50%, 100%, or even 500% slower (depending on where the term in question appears in hint patterns and goal terms). 

Obviously one can make `reverse_coercion` TC opaque but then we might see failing TC searches, especially in databases with tight opacity settings.

**Comment by andres-erbsen** [comment]
@proux01, thank you for the explanation. My concerns are about keeping the effort required to maintain and write reusable tactics low, not backwards compatibility in the strict sense of whether old code will keep compiling. I understand that code from before reversible coercions were released would not be broken by this change. However, it seems that from a tactic-maintainer's perspective, `reverse_coercion` might as well be a part of the term AST: unsuspecting users of the tactic would read the Coq reference manual and write terms that appear distinct to the tactic. Further, leaving ` reverse_coercion` uninterpreted is wrong from the perspective of (almost?) all tactics: it instead should be ignored, but not unfolded so that it remains in the output.

The specific example with Fiat Cryptography is also of this form: Jason Gross, Samuel Gruetter, and I maintain a large library of Ltac&Ltac2 code that is used by less experienced but independent and competent engineers, and our tactics not working with code that they wrote following the Coq reference manual would be unwelcome friction. Examples of this kind of issues that we already are running into include primitive vs nonprimitive projections and unipoly annotations. Porting tactics to handle new features in the elaborator's output language is evidently costly: many standard-library tactics have not been updated for those two cases yet, and even those who have been took a while. Thus, I think we should at least consider alternative strategies to achieve the printing improvements in this PR.

**Comment by proux01** [comment]
> However, it seems that from a tactic-maintainer's perspective, `reverse_coercion` might as well be a part of the term AST: 

Right, but its an elaborator think, no need to make it into the kernel.

> unsuspecting users of the tactic would read the Coq reference manual and write terms that appear distinct to the tactic.

That's the whole point of having an elaborator.

> Further, leaving `reverse_coercion` uninterpreted is wrong from the perspective of (almost?) all tactics: it instead should be ignored, but not unfolded so that it remains in the output.

I'm not sure I understand this? What do you mean by "leaving `reverse_coercion` uninterpreted"? What do you mean by "it instead should be ignored, but not unfolded so that it remains in the output."? If you unfold `reverse_coercion` it will basically disappear since it is just an elaborated identity.
 
> I think we should at least consider alternative strategies to achieve the printing improvements in this PR.

What alternative strategies do you offer?

**Comment by Alizter** [comment]
By the way, I am not particularly following the issue here. Does anybody have a problematic example in mind?

**Comment by andres-erbsen** [comment]
Concrete example:

```coq
#[universes(polymorphic=yes)] Definition ReverseCoercionSource (T : Type) := T.
#[universes(polymorphic=yes)] Definition ReverseCoercionTarget (T : Type) := T.
#[nonuniform, reversible=no, universes(polymorphic=yes)]
Coercion reverse_coercion {T' T} (x' : T') (x : ReverseCoercionSource T)
  : ReverseCoercionTarget T' := x'.
Register reverse_coercion as core.coercion.reverse_coercion.
Structure S := { ssort :> Type; sstuff : ssort }.
Canonical Structure S_nat := {| ssort := nat; sstuff := 0; |}.
Goal True.
  constr_eq constr:(S_nat) constr:(nat : S).
  constr_eq constr:(S_nat) constr:(@reverse_coercion S Set S_nat nat : S). (* fails *)

  assert (S_nat = (nat : S)) by congruence.
  assert (S_nat = (@reverse_coercion S Set S_nat nat : S)) by congruence. (* fails *)

  (* and so on, with more complicated tactic interactions*)
Abort.
```

I will think about what else we would be able to do. One option might be to have a notion of "printing hints", rules of the form "if you see a term that matches this pattern instead print the right-hand side of the pattern".

**Comment by gares** [comment]
@andres-erbsen I don't get this line of argument, you seem to give examples of why syntactic equality is a sub relation of conversion:
```
  constr_eq constr:(0) constr:(id 0). 
```
the real test is, IMO:
```
  constr_eq constr:(nat) constr:(nat : S).
```
which is already false.

Of course, if you had code doing
```
  constr_eq constr:(S_nat) constr:(nat : S).
```
in 8.17, well that would break in 8.18 with this PR merged.
But you don't have it, so it is not concrete to me.

**Comment by andres-erbsen** [comment]
@gares: maybe the difference between our contexts is that unification is almost always too slow to be used in my tactics, so we rely on syntactic equality a lot, hence the elaborator breaking it would be a problem.

I think the goal here is that `Check S_nat` should print `S_nat`; `S_nat` should only be printed as `nat` in contexts where `nat` would be elaborated to `S_nat`. A solution to achieve this without being sensitive to how the term was originally input would seem much more desirable to me.

**Review by tabareau — APPROVED** [review]
We want this kind of features to remain outside the kernel and TCB so the current technical choice seems the more appropriate. This still leaves space for improvements and we invite users to report issues or propose CEP if they find points of amelioration.

**Comment by andres-erbsen** [comment]
I don't think anyone is arguing that a solution to the problem this PR is trying to address would belong in the kernel. That said, I am not sure what other commenters hope would happen, and I get the sense that I may not be the only one, so I will try to be very clear about how I understand the situation.

My position is just that the current implementation is a step in the wrong direction, and we should seek an essentially different non-kernel solution. If this PR is merged, all tactics would need to be updated to handle the ad-hoc encoding of the  "trace of what the user actually did"  (the new definition introduced in this PR). If not, these tactics would be effectively broken to an unsuspecting user of any library that uses reversible coercions. This PR makes no attempt to update the tactics to restore the level of predictability enjoyed with other coercions (or with reversible coercions as they stand in 8.17). Indeed, it is not clear what these adaptations would look like, and IMO that is a clear sign that the design space needs to be considered more broadly before proceeding. Similar messes left behind with the addition of primitive projections and universe polymorphism continue to be a bane to users and developers.

The direction I think might be fruitful is to try to find a notion of context-dependent printing that is more fine-grained than scopes (and perhaps type-based). In this case, I think the desired context-based rule would involve approximating an answer to the question "would nat in this context be elaborated as `S_nat` without relying on any marker definitions inserted by elaboration. I agree that figuring out how to make this precise would likely require enough work for a CEP and some high-bandwidth brainstorming sessions.

**Comment by gares** [comment]
We may have found an alternative path, which does not require touching the kernel (as other proposals here). It is fully at the printer level, but requires to be implemented and tested for performance

**Comment by proux01** [comment]
I tried to implement the idea discussed in https://coq.zulipchat.com/#narrow/stream/237656-Coq-devs-.26-plugin-devs/topic/Only.20printing.20abbreviations it looks nice when canonical structures have no argument but I fear getting it to work with arguments would be way too heuristic hence probably constantly broken.

**Comment by proux01** [comment]
@coqbot run full ci

**Comment by tabareau** [comment]
@coqbot merge now

**Comment by andres-erbsen** [comment]
What happened here? (Right until the merge I was under the impression that this was blocked on finding a satisfactory story for tactic compatibility, but it looks like it got merged without that?)

**Comment by Janno** [comment]
I am equally confused. Was there private discussion that isn't reflected in this MR? It would be nice to understand the reasoning.

**Comment by Janno** [comment]
Seeing https://github.com/coq/coq/pull/17484 I am now wondering if the new term will result in more difficulties in tactics that construct terms unsafely. It seems that the application of `reverse_coercion` to its arguments is typechecked automatically but is there any guarantee that the resulting universe constraints are sufficient to typecheck `reverse_coercion` in the context into which it is inserted (assuming it would have typechecked without the added `reverse_coercion` term)?

If this is guaranteed, why did HB need changes?

Also, given that every inserted `reverse_coercion` term gets fresh universes should this MR have had a benchmark? There could be Qed performance regressions. (Although I am not sure we would catch them given how reverse coercions in general are relatively new and probably not often used.)

**Comment by proux01** [comment]
> If this is guaranteed, why did HB need changes?

HB was designed at a time coq-elpi had no support for universe polymorphism so this kind of bug was somewhat expected.

> Also, given that every inserted `reverse_coercion` term gets fresh universes should this MR have had a benchmark? There could be Qed performance regressions. (Although I am not sure we would catch them given how reverse coercions in general are relatively new and probably not often used.)

You have your answer, reverse coercions are too recent to get enough use in the CI codebase for a benchmark to be of any interest.

**Comment by gares** [comment]
> If this is guaranteed, why did HB need changes?

Since Coq 8.5 a term which is well typed can be rejected by the kernel if it is not accompanied by its universe constraints (even if the kernel is able to infer the missing ones, a bad choice IMO, but that is OT). Tthat explains the bug:
the code in HB was wrongly passing to the kernel well typed terms without all the universe constraints, the extra calls to coq.typecheck are amending that (typecheck is not the kernel typecheck, it is the pretyper one which happily infers the constraint and, in the elpi API, stores then in the current state).

In some sense this patch exposed a bug which was already there (w.r.t. the spec of Coq's kernel).

**Comment by Janno** [comment]
I am aware of the kernel's requirements. In the kind of automation I write I usually insert strategic calls to typechecking to generate exactly the constraints that are necessary to make the kernel accept the full proof term. In many cases this boils down to typechecking a partial application to which the remaining arguments are then added unsafely. But most of the time we can bypass typechecking entirely since we know exactly that all the constraints are already established by some other mechanism. 

I was wondering if I should except breakage from schemes like that due to `reverse_coercion` possibly showing up with fresh universes.

**Comment by gares** [comment]
Given your code does not use the feature of reverse coercions, I think you should not see any change.

**Pull request merged**
